File: src/website/services/p2p/addOffer.ts
import { Option } from "../../domain/model/common";
import { P2PConnectionOffer } from "../../domain/model/p2pConnectionDetails";
import { UserId } from "../../domain/model/users";
import { fromCandid as optionFromCandid } from "../candidConverters/option";
import { toCandid as userIdToCandid } from "../candidConverters/userId";
import CanisterClientFactory from "../CanisterClientFactory";

export default async function(request: AddOfferRequest) : Promise<AddOfferResponse> {
    const client = CanisterClientFactory.current!.p2pClient;
    const canisterRequest = {
        id: request.id,
        user_id: userIdToCandid(request.userId),
        connection_string: request.connectionString,
        ice_candidates: request.iceCandidates
    };
    const response = await client.add_offer(canisterRequest);

    if (response.hasOwnProperty("Success")) {
        const result = response.Success;
        const candidCounterOffer: any = optionFromCandid(result.existing_counter_offer);
        const counterOffer: Option<P2PConnectionOffer> = candidCounterOffer
            ? {
                kind: "offer",
                id: candidCounterOffer.id,
                userId: request.userId,
                connectionString: candidCounterOffer.connection_string,
                iceCandidates: candidCounterOffer.ice_candidates,
                ageSeconds: candidCounterOffer.age_seconds
            }
            : null;

        return {
            kind: "success",
            offerAdded: result.offer_added,
            existingCounterOffer: counterOffer
        };
    } else {
        throw new Error("Unrecognised 'add_offer' response");
    }
}

export type AddOfferRequest = {
    id: string,
    userId: UserId,
    connectionString: string,
    iceCandidates: string[]
}

export type AddOfferResponse =
    Success;

export type Success = {
    kind: "success",
    offerAdded: boolean,
    existingCounterOffer: Option<P2PConnectionOffer>
}

File: src/services/p2p/src/updates/add_offer.rs
use ic_cdk::export::candid::CandidType;
use ic_cdk::storage;
use serde::Deserialize;
use shared::user_id::UserId;
use crate::domain::connection_details::{AllConnectionDetails, CounterOffer};

pub fn update(request: Request) -> Response {
    let me = shared::user_id::get_current();
    let now = shared::timestamp::now();
    let connection_details: &mut AllConnectionDetails = storage::get_mut();

    let counter_offer = connection_details.add_offer(
        request.id,
        me,
        request.user_id,
        request.connection_string,
        request.ice_candidates,
        now);

    let result = Result {
        offer_added: counter_offer.is_none(),
        existing_counter_offer: counter_offer
    };

    Response::Success(result)
}

#[derive(Deserialize)]
pub struct Request {
    id: String,
    user_id: UserId,
    connection_string: String,
    ice_candidates: Vec<String>
}

#[derive(CandidType)]
pub enum Response {
    Success(Result)
}

#[derive(CandidType)]
pub struct Result {
    offer_added: bool,
    existing_counter_offer: Option<CounterOffer>
}
